conda_gitenv/resolve.py

```diff
@@ -12,46 +12,68 @@
 
 import datetime
 import contextlib
+from fnmatch import fnmatch
 import logging
 import os
 import shutil
 import tempfile
 import warnings
 
-import conda.resolve
 import conda.api
+import conda.resolve
 import conda_build_all.version_matrix
 from git import Repo
 import yaml
 
 from conda_gitenv import manifest_branch_prefix
 
 
-def resolve_spec(spec_fh):
+def resolve_spec(spec_fh, api_user, api_key):
     """
-    Given an open file handle to an env.spec, return a list of strings containing
-    '<channel_url>\t<pkg_name>' for each package resolved.
+    Given an open file handle to an env.spec, return a list of strings
+    containing '<channel_url>\t<pkg_name>' for each package resolved.
 
     """
+    try:
+        # Python3...
+        from urllib.parse import urlparse
+    except ImportError:
+        # Python2...
+        from urlparse import urlparse
+
     spec = yaml.safe_load(spec_fh)
     env_spec = spec.get('env', [])
-    index = conda.api.get_index(spec.get('channels', []), prepend=False, use_cache=False)
+    channels = spec.get('channels', [])
+
+    # Inject the API user and key into the channel URLs...
+    if api_user and api_key:
+        for i, url in enumerate(channels):
+            parts = urlparse(url)
+            api_url = '{}://{}:{}@{}{}'.format(parts.scheme, api_user, api_key,
+                                               parts.netloc, parts.path)
+            channels[i] = api_url
+
+    index = conda.api.get_index(channels, prepend=False, use_cache=False)
     resolver = conda.resolve.Resolve(index)
     packages = sorted(resolver.solve(env_spec),
                       key=lambda pkg: pkg.dist_name.lower())
 
     pkgs = []
     for pkg in packages:
         pkg_info = index[pkg]
-        pkgs.append('\t'.join([pkg_info['channel'],
-                               pkg_info['fn'][:-len('.tar.bz2')]])), 
+        pkgs.append('\t'.join([os.path.join(pkg_info['schannel'],
+                                            pkg_info['subdir']),
+                               pkg_info['fn'][:-len('.tar.bz2')]]))
```

Comment:
I'm guessing using `schannel` and `subdir` ensures we don't expose the API key, which `channel` would do?

Reply:
@lbdreyer Yup, you're on the money :+1: 

For example, previously we'd join:
- `channel` = `https://channel/linux-64` and
- `fn` = `package.tar.bz2`
to give:
- `https://channel/linux-64/package`

The `fn` member of the `IndexRecord` hasn't changed, but when we inject the API user and key, the `channel` is correctly polluted with those details i.e., `https://api-user:api-key@channel/linux-64`, which we don't want to expose.

We can use the `schannel` and `subdir` to easily do this. In this example, `schannel` is simply `https://channel` and `subdir` is `linux-64`, which we use along with `fn` to construct the filtered package information `https://channel/linux-64/package`.